File: sdk/javascript/secureSignClient.js
/**
 * SecureSign JavaScript SDK (referencia) — cliente delgado sobre la API REST
 * documentada en docs/05-integracion/manual-integracion-api.md.
 *
 * Uso previsto: backend Node.js de un sistema integrador. NUNCA usar en
 * frontend con el apiKey embebido — ver manual, capítulo 7 (buenas prácticas).
 */
class SecureSignClient {
  constructor({ apiKey, tenant, baseUrl = "https://api.securesign.pe/v1" }) {
    if (!apiKey) throw new Error("apiKey es requerido");
    this.apiKey = apiKey;
    this.tenant = tenant;
    this.baseUrl = baseUrl;
    this._accessToken = null;
    this._expiraEn = 0;

    this.documentos = new DocumentosResource(this);
    this.firmas = new FirmasResource(this);
  }

  async _obtenerToken() {
    if (this._accessToken && Date.now() < this._expiraEn) return this._accessToken;

    const respuesta = await fetch(`${this.baseUrl}/auth/token`, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({ grant_type: "client_credentials", client_id: this.tenant, client_secret: this.apiKey }),
    });

    if (!respuesta.ok) throw new Error(`No se pudo autenticar: ${respuesta.status}`);

    const datos = await respuesta.json();
    this._accessToken = datos.access_token;
    this._expiraEn = Date.now() + (datos.expires_in - 30) * 1000; // margen de 30s
    return this._accessToken;
  }

  async _request(path, options = {}) {
    const token = await this._obtenerToken();
    const respuesta = await fetch(`${this.baseUrl}${path}`, {
      ...options,
      headers: { Authorization: `Bearer ${token}`, ...options.headers },
    });

    if (!respuesta.ok) {
      const error = await respuesta.json().catch(() => ({}));
      throw new Error(error.mensaje || `Error ${respuesta.status} en ${path}`);
    }
    return respuesta.json();
  }
}

class DocumentosResource {
  constructor(client) { this.client = client; }

  async registrar({ archivo, codigoExterno, usuarioSolicitanteId }) {
    const formData = new FormData();
    formData.append("archivo", archivo);
    if (codigoExterno) formData.append("codigoExterno", codigoExterno);
    if (usuarioSolicitanteId) formData.append("usuarioSolicitanteId", usuarioSolicitanteId);

    return this.client._request("/documentos", { method: "POST", body: formData });
  }

  async descargarFirmado(idDocumento) {
    return this.client._request(`/documentos/${idDocumento}/firmado`);
  }
}

class FirmasResource {
  constructor(client) { this.client = client; }

  async crearSolicitud({ documentoId, tipoFirma, requiereOrdenSecuencial = false, firmantes, fechaLimite }) {
    return this.client._request("/firmas/solicitudes", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ documentoId, tipoFirma, requiereOrdenSecuencial, firmantes, fechaLimite }),
    });
  }

  async consultarEstado(solicitudFirmaId) {
    return this.client._request(`/firmas/${solicitudFirmaId}/estado`);
  }
}

module.exports = { SecureSignClient };

File: sdk/python/securesign_client.py
"""
SecureSign Python SDK (referencia) — cliente delgado sobre la API REST
documentada en docs/05-integracion/manual-integracion-api.md.

Requiere: pip install requests
"""
import time
from dataclasses import dataclass
from typing import Optional

import requests


@dataclass
class SolicitudFirmaResponse:
    solicitud_firma_id: str
    codigo_verificacion_publico: str
    url_firma: str


class SecureSignClient:
    def __init__(self, api_key: str, tenant: str, base_url: str = "https://api.securesign.pe/v1"):
        self.api_key = api_key
        self.tenant = tenant
        self.base_url = base_url
        self._access_token: Optional[str] = None
        self._expira_en: float = 0

        self.documentos = _DocumentosResource(self)
        self.firmas = _FirmasResource(self)

    def _obtener_token(self) -> str:
        if self._access_token and time.time() < self._expira_en:
            return self._access_token

        respuesta = requests.post(
            f"{self.base_url}/auth/token",
            data={"grant_type": "client_credentials", "client_id": self.tenant, "client_secret": self.api_key},
        )
        respuesta.raise_for_status()
        datos = respuesta.json()
        self._access_token = datos["access_token"]
        self._expira_en = time.time() + datos["expires_in"] - 30
        return self._access_token

    def _request(self, method: str, path: str, **kwargs) -> dict:
        token = self._obtener_token()
        headers = kwargs.pop("headers", {})
        headers["Authorization"] = f"Bearer {token}"
        respuesta = requests.request(method, f"{self.base_url}{path}", headers=headers, **kwargs)
        respuesta.raise_for_status()
        return respuesta.json()


class _DocumentosResource:
    def __init__(self, client: SecureSignClient):
        self._client = client

    def registrar(self, archivo: str, codigo_externo: Optional[str] = None, usuario_solicitante_id: Optional[str] = None) -> dict:
        with open(archivo, "rb") as f:
            files = {"archivo": f}
            data = {}
            if codigo_externo:
                data["codigoExterno"] = codigo_externo
            if usuario_solicitante_id:
                data["usuarioSolicitanteId"] = usuario_solicitante_id
            return self._client._request("POST", "/documentos", files=files, data=data)

    def descargar_firmado(self, id_documento: str) -> dict:
        return self._client._request("GET", f"/documentos/{id_documento}/firmado")


class _FirmasResource:
    def __init__(self, client: SecureSignClient):
        self._client = client

    def crear_solicitud(self, documento_id: str, tipo_firma: str, firmantes: list, requiere_orden_secuencial: bool = False, fecha_limite: Optional[str] = None) -> dict:
        payload = {
            "documentoId": documento_id,
            "tipoFirma": tipo_firma,
            "requiereOrdenSecuencial": requiere_orden_secuencial,
            "firmantes": firmantes,
            "fechaLimite": fecha_limite,
        }
        return self._client._request("POST", "/firmas/solicitudes", json=payload)

    def consultar_estado(self, solicitud_firma_id: str) -> dict:
        return self._client._request("GET", f"/firmas/{solicitud_firma_id}/estado")
